cdc/sinkv2/eventsink/cloudstorage/dml_worker.go

```diff
@@ -0,0 +1,304 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package cloudstorage
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/pingcap/log"
+	"github.com/pingcap/tidb/br/pkg/storage"
+	"github.com/pingcap/tiflow/cdc/model"
+	"github.com/pingcap/tiflow/cdc/sink/codec/common"
+	"github.com/pingcap/tiflow/pkg/chann"
+	"go.uber.org/zap"
+)
+
+// dmlWorker denotes a worker responsible for writing messages to cloud storage.
+type dmlWorker struct {
+	// worker id
+	id            int
+	changeFeedID  model.ChangeFeedID
+	storage       storage.ExternalStorage
+	flushInterval time.Duration
+	// msgChannels maintains a mapping of <table, unbounded message channel>.
+	msgChannels sync.Map
+	// flushNotifyCh is used to notify that several tables can be flushed.
+	flushNotifyCh chan flushTask
+	// defragmenters maintains a mapping of <table, defragmenter>.
+	defragmenters sync.Map
+	// fileIndex maintains a mapping of <table, index number>.
+	fileIndex map[versionedTable]uint64
+	// activeTablesCh is used to notify that a group of eventFragments are
+	// sent to the corresponding defragmenter and we can reap messages from
+	// the defragmenter right now.
+	activeTablesCh *chann.Chann[versionedTable]
+	wg             sync.WaitGroup
+	isClosed       uint64
+	errCh          chan<- error
+	extension      string
+}
+
+// flushTask defines a task containing the tables to be flushed.
+type flushTask struct {
+	activeTables []versionedTable
+}
+
+func newDMLWorker(
+	id int,
+	changefeedID model.ChangeFeedID,
+	storage storage.ExternalStorage,
+	flushInterval time.Duration,
+	extension string,
+	errCh chan<- error,
+) *dmlWorker {
+	d := &dmlWorker{
+		id:             id,
+		changeFeedID:   changefeedID,
+		storage:        storage,
+		flushInterval:  flushInterval,
+		flushNotifyCh:  make(chan flushTask, 1),
+		activeTablesCh: chann.New[versionedTable](),
+		fileIndex:      make(map[versionedTable]uint64),
+		extension:      extension,
+		errCh:          errCh,
+	}
+
+	return d
+}
+
+// run creates a set of background goroutines.
+func (d *dmlWorker) run(ctx context.Context, ch *chann.Chann[eventFragment]) {
+	d.backgroundDefragmentMsgs(ctx)
```

Comment:
I think it's possible to remove this goroutine, if we orgnize txns based on table, and then schedule tasks to uploader also based on table.

Reply:
This `backgroundDefragmentMsgs` goroutine is removed in the latest commit.